src/bt.rs

```diff
@@ -0,0 +1,423 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! An append-only binary tree.
+//!
+//! ## Properties:
+//! - Binary tree: nodes have either 0, 1, or 2 child nodes.
+//! - Append-only: tree only grows, values can only be inserted and queried.
+//! - Serializable: tree can be coverted to and from bytes.
+//!
+//! ## Creation
+//! Use [`BinaryTree::default`] to create a binary tree, note that there must be
+//! one value at the root node.
+//!
+//! ## Insertion
+//! Given a value and a binary path, [`BinaryTree::insert`] stores the value
+//! at the end of the path.
+//!
+//! ## Query
+//! Given a binary path, [`BinaryTree::get`] returns a reference to the value
+//! stored in the node located at the end of the path. The empty path returns
+//! the value at the root node. No value is returned if the end of the path
+//! is unreachable.
+//!
+//! ## Serialization
+//! A tree can be serializad to and from bytes using [`BinaryTree::encode`]
+//! and [`BinaryTree::decode_with_param`] functions, respectively. One-byte
+//! markers are used allowing to store sparse trees with a lower overhead.
+//!
+//! ## Example
+//! This binary tree can be created with the following code:
+//!
+//! ```txt
+//!               ()=@
+//!          /              \
+//!       (0)=a           (1)=b
+//!     /      \         /      \
+//! (00)=c  (01)=d   (10)=e  (11)=f
+//! ```
+//!
+//! use prio::bt::BinaryTree;
+//! use bitvec::{bits,prelude::Lsb0};
+//! let mut tree = BinaryTree::default();
+//! tree.insert(bits!(), '@');
+//! tree.insert(bits!(0), 'a');
+//! tree.insert(bits!(1), 'b');
+//! tree.insert(bits!(0, 0), 'c');
+//! tree.insert(bits!(0, 1), 'd');
+//! tree.insert(bits!(1, 0), 'e');
+//! tree.insert(bits!(1, 1), 'f');
+
+use core::fmt::Debug;
+use std::io::Cursor;
+
+use bitvec::slice::BitSlice;
+
+use crate::codec::{CodecError, Decode, Encode, ParameterizedDecode};
+
+/// Errors triggered by binary tree operations.
+#[derive(Debug, thiserror::Error)]
+#[non_exhaustive]
+pub enum BinaryTreeError<V> {
+    /// Error when inserting in a node with existing value.
+    #[error("node already contains a value")]
+    InsertNonEmptyNode(V),
+    /// Error when an operation cannot reach the node at the end of a path.
+    #[error("unreachable node in the tree")]
+    UnreachableNode(V),
+}
+
+/// Used to indicate a traversal path on the binary tree.
+pub type Path = BitSlice;
+
+type SubTree<V> = Option<Box<Node<V>>>;
+
+/// Represents a node of a binary tree.
+pub struct Node<V> {
+    value: V,
+    left: SubTree<V>,
+    right: SubTree<V>,
+}
```

Comment:
This seems like the most natural way to implement a binary tree in Rust, using `Box` to emulate pointers in C/Go/etc. However, I wonder if it's the most efficient thing we could do.

Alternatively, we could keep the tree in a a vector of nodes and use indices into the vector for the children:

```rust
pub struct Node<V> {
    value: V,
    left: usize,
    right: usize,
}

pub struct Tree<V>(Vec<Node<V>>);
```

This would make the API a bit messier (when traversing the tree from one node to another, you need to have a reference to the tree), but this messiness would be crate-private. The benefit, I suspect, would be fewer allocations. I have no idea if this would be worth it however.

Thoughts, @divergentdave?

Reply:
That's a good idea, however, I would do it in a different PR, so one can contrast timings.

Reply:
Drive-by:

If the tree is unlikely to be sparse (or if the tree is sparse, but an acceptable sparse-vector implementation is available, or perhaps even a map from `usize` to `V`), the representation could be:

```rust
pub struct Tree<V>(Vec<Option<V>>);
/* or */
pub struct Tree<V>(HashMap<usize, Option<V>>);
```

... using an implicit rule to determine the children of a node: the value for the root node is stored at index 0; the children of the node whose value is stored at index `i` have their values stored at indices `2*i + 1` and `2*i + 2`.

This maintains the likely cache-locality/lowered-allocation wins of the node-referenced-by-index tree, while permitting fewer bad tree representations (i.e. it would not be possible to construct a "tree" with a cycle in it), and slightly simplifying tree traversal.

Insertion could _almost_ be reduced to a simple index-lookup-and-set based on an index that has the same bitwise representation as the path being inserted to, except that this tree implementation currently requires all intermediate nodes to have a value. So insertion would need to check all of the intermediate nodes as well; if that requirement was relaxed, the need to check intermediate nodes could be dropped.

Reply:
Yes, I expect that using one big allocation as an offset-indexed arena of nodes could improve performance by improving cache locality, and amortizing malloc calls. It would be helpful if we could use heuristics from the application to size the `Vec`, to avoid unnecessary copying when resizing, but `Decode` doesn't lend itself to this.

For practical reasons, this tree will need to be sparse, as we could easily have a tree depth of over 100.

Reply:
@armfazh  proposes to land this PR as-is and consider this change later on. I'm cool with that if you all are.